### DISCUSSION TITLE
Parallel run/parallel worker identifier?

### DISCUSSION THREAD

**Discussion by bbascarevic** [xunit/xunit#2496] [Question] [answered]
Is there some sort of a parallel run/parallel worker identifier available in Xunit? String, integer, anything will do as long as it's unique per parallel run.

**Comment by bartelink** [comment, marked as answer]
Nothing built in - `Guid.NewGuid()`, or a `static int` with an `Interlocked.Increment` should do the trick

**Reply by bbascarevic** [reply]
@bartelink  I fail to understand how any of that would help. 

The runner decides how many parallel workers will it spawn, based on a parameter, number of physical processors, or whatever. Tests are then distributed among those workers and executed in parallel. I need to access the identifier of the current worker that's executing the test from within the test, at runtime. All tests executed by one distinct worker must have one distinct identifier. The identifier can be a simple ordinal of the worker, or a UUID, or whatever, as long as it's the same for all tests executed by that worker. 

How is parallelization and isolation achieved in Xunit - threads, processes, appdomains? If it's processes I could use the process ID or an environment variable.

**Reply by bartelink** [reply]
I have not studied the impl but I believe the reason there is nothing of this kind in XUnit is down to:
a) it's something that can change, either to make the impl work better, or to make it operate in more environments
b) In general the impl is not a case of making N threads and they run until the end - think of it instead as a semaphore that guards against too many instances

The runner is ultimately responsible for starting as many tests as the constraints on them make possible, and replacing any that have completed with new ones, but that can be done via the thread pool and many other mechanisms (and is something that has necessarily varied over time)

(aside: AppDomains are specific to .NET old/full framework)

Perhaps it might be better for you to describe your aims rather than a specific thing, and others may be able to help, i.e. https://xyproblem.info

**Reply by bbascarevic** [reply]
I need to write integration tests in Xunit. The system under test connects to real instances of dependent services during integration testing and I need to make sure that parallel test workers don't step on each others' toes and cause flaky tests. The way to do that is to prepare independent instances of dependent services, one for each parallel worker. The cost of establishing an environment per test (as opposed to per parallel worker) would be too high and also unnecessary. Dependent services are capable of resetting their state via an API (which is very fast) but the time cost of spinning them up is relatively high.

**Reply by bartelink** [reply]
Cool; that's a question I can work with ;)

Collection Fixtures can sometimes be relevant in that space, and are actively managed by Xunit with the intention of providing something that is not reliant on low level impl details, though it has to be noted that the authors regularly mention that supporting integration testing is not a goal of Xunit, so the experience may not be optimal.

The bottom line for your scenario is that there is no big assumption or tool you can rely on; the main tool at your disposal is
1. obtain your resources before/during the test (either store a value or use a `Lazy` local to the test)
2. hand them back after (implement IDisposable.Dispose)

In other words, a Rent/Return model kinda like how object pools are used to reduce GC pressure.

The other thing to note is that a Class Fixture can be relevant, though probably not for your scenario - it lets you have some state that's shared/recycled amount tests within a single Test Class.

**Reply by bbascarevic** [reply]
@bartelink I thought of that as well, but I figured that a pool is only useful if all "workers" use the same pool, which can really only be the case if there's a single test process that hosts threads that compete for pending tests (again, I don't know how parallelization is achieved under the hood).

**Reply by bartelink** [reply]
Parallelization is achieved by Xunit via the thread pool. While there are ways to involve multiple processes, the default/typical run config is single-process (otherwise parallelization is not limited to whats a reasonable amount of processes to have in flight - as with having a thread per test, having a process per test would induce a high cost)

If you implement a pool, then you can have it dictate a maximum number of state instances that you will recycle. Xunit will parallelize based on its constraints - the main ones of those are that there's one test in flight per Test Class (or Test Collection where a Test Class uses a Collection Fixture)
